apollo-router/src/plugins/external.rs

```diff
@@ -12,41 +11,172 @@ use bytes::Bytes;
+
+        let http_client = ServiceBuilder::new()
+            .layer(TimeoutLayer::new(init.config.timeout))
+            .service(hyper::Client::builder().build(connector));
+
+        ExternalPlugin::new(http_client, init.config, init.supergraph_sdl)
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        self.router_service(service)
+    }
+
+    fn supergraph_service(&self, service: supergraph::BoxService) -> supergraph::BoxService {
+        self.supergraph_service(service)
+    }
+
+    fn execution_service(&self, service: execution::BoxService) -> execution::BoxService {
+        self.execution_service(service)
+    }
+
+    fn subgraph_service(&self, name: &str, service: subgraph::BoxService) -> subgraph::BoxService {
+        self.subgraph_service(name, service)
+    }
+}
+
+// This macro allows us to use it in our plugin registry!
+// register_plugin takes a group name, and a plugin name.
+//
+// In order to keep the plugin names consistent,
+// we use using the `Reverse domain name notation`
+register_plugin!(
+    "experimental",
+    "external",
+    ExternalPlugin<HTTPClientService>
+);
+
+// -------------------------------------------------------------------------------------------------------
+
+/// This is where the real implementation happens.
+/// The structure above calls the functions defined below.
+///
+/// This structure is generic over the HTTP Service so we can test the plugin seamlessly.
 #[derive(Debug)]
-struct ExternalPlugin {
+struct ExternalPlugin<C>
+where
+    C: Service<hyper::Request<Body>, Response = hyper::Response<Body>, Error = BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <C as tower::Service<http::Request<hyper::Body>>>::Future: Send + Sync + 'static,
+{
+    http_client: C,
     configuration: Conf,
     sdl: Arc<String>,
 }
 
-/// What information is passed to a request/response stage
+impl<C> ExternalPlugin<C>
+where
+    C: Service<hyper::Request<Body>, Response = hyper::Response<Body>, Error = BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <C as tower::Service<http::Request<hyper::Body>>>::Future: Send + Sync + 'static,
+{
+    fn new(http_client: C, configuration: Conf, sdl: Arc<String>) -> Result<Self, BoxError> {
+        Ok(Self {
+            http_client,
+            configuration,
+            sdl,
+        })
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        self.configuration.stages.router.as_service(
+            self.http_client.clone(),
+            service,
+            self.configuration.url.clone(),
+            self.sdl.clone(),
+        )
+    }
+
+    fn supergraph_service(&self, service: supergraph::BoxService) -> supergraph::BoxService {
+        service
+    }
+
+    fn execution_service(&self, service: execution::BoxService) -> execution::BoxService {
+        service
+    }
+
```

Comment:
```suggestion
```

Reply:
fixed in aa0ec75d27d5eae6cdee7a3f2b5f922ee26ee785